docs/tutorial/native-code-and-electron-cpp-linux.md

```diff
@@ -0,0 +1,1595 @@
+          {
+            delete callbackData;
+            return;
+          }
+
+          try
+          {
+            auto callback = addon->callbacks.Value().Get(callbackData->eventType).As<Napi::Function>();
+            if (callback.IsFunction())
+            {
+              callback.Call(addon->emitter.Value(), {Napi::String::New(napi_env, callbackData->payload)});
+            }
+          }
+          catch (...)
+          {
+          }
+
+          delete callbackData;
+        },
+        &tsfn_);
+
+    if (status != napi_ok)
+    {
+      Napi::Error::New(env_, "Failed to create threadsafe function").ThrowAsJavaScriptException();
+      return;
+    }
+
+    // Set up the callbacks here
+    auto makeCallback = [this](const std::string &eventType)
+    {
+      return [this, eventType](const std::string &payload)
+      {
+        if (tsfn_ != nullptr)
+        {
+          auto *data = new CallbackData{
+              eventType,
+              payload,
+              this};
+          napi_call_threadsafe_function(tsfn_, data, napi_tsfn_blocking);
+        }
+      };
+    };
+
+    cpp_code::setTodoAddedCallback(makeCallback("todoAdded"));
+    cpp_code::setTodoUpdatedCallback(makeCallback("todoUpdated"));
+    cpp_code::setTodoDeletedCallback(makeCallback("todoDeleted"));
+  }
+```
+
+This is the most complex part of our bridge, implementing bidirectional communication. There are a few things worth noting going on here, so let's take them step by step:
+
+`CallbackData` struct:
+
+* Holds the event type, JSON payload, and a reference to our addon.
+
+In the constructor:
+
+* We create a thread-safe function (`napi_create_threadsafe_function`) which is crucial for calling into JavaScript from the GTK3 thread
+* The thread-safe function callback unpacks the data and calls the appropriate JavaScript callback
+* We create a lambda `makeCallback` that produces callback functions for different event types
+* We register these callbacks with our C++ code using the setter functions
+
+Let's talk about `napi_create_threadsafe_function`. The orchestration of different threads is maybe the most difficult part about native addon development - and in our experience, the place where developers are most likely to give up. `napi_create_threadsafe_function` is provided by the N-API and allows you to safely call JavaScript functions from any thread. This is essential when working with GUI frameworks like GTK3 that run on their own thread. Here's why it's important:
+
+1. **Thread Safety**: JavaScript in Electron runs on a single thread (exceptions apply, but this is a generally useful rule). Without thread-safe functions, calling JavaScript from another thread would cause crashes or race conditions.
+2. **Queue Management**: It automatically queues function calls and executes them on the JavaScript thread.
+3. **Resource Management**: It handles proper reference counting to ensure objects aren't garbage collected while still needed.
```

Comment:
```suggestion
1. **Thread Safety**: JavaScript in Electron runs on a single thread (exceptions apply, but this is a generally useful rule). Without thread-safe functions, calling JavaScript from another thread would cause crashes or race conditions.
1. **Queue Management**: It automatically queues function calls and executes them on the JavaScript thread.
1. **Resource Management**: It handles proper reference counting to ensure objects aren't garbage collected while still needed.
```